File: homework/Manh_Nguyen_Van/lesson-1/304-range-sum-query-2d-immutable.java
class NumMatrix {

    int[][] sumMatrix;
    
    public NumMatrix(int[][] matrix) {
        
        sumMatrix = new int[matrix.length + 1][matrix[0].length + 1];
        for (int i = 1; i < matrix.length + 1; i++) {
            for (int j = 1; j < matrix[0].length + 1; j++) {
                sumMatrix[i][j] = sumMatrix[i][j - 1] + sumMatrix[i - 1][j] - sumMatrix[i - 1][j - 1]  + matrix[i - 1][j - 1];
            }
        }
        printMatrix(sumMatrix);
    }
    
    public int sumRegion(int row1, int col1, int row2, int col2) {
        return sumMatrix[row2+1][col2+1] - sumMatrix[row1][col2+1] - sumMatrix[row2+1][col1] + sumMatrix[row1][col1];
    }
	
	private static void printArr(int[] nums) {
		System.out.print("[");
		for (int i = 0; i < nums.length; i++) {
			System.out.print(String.format(" %d ", nums[i]));
		}
		System.out.println("]");
	}

	private static void printMatrix(int[][] matrix) {
		System.out.println("[");
		for (int i = 0; i < matrix.length; i++) {
			printArr(matrix[i]);
		}
		System.out.println("]");
	}
}

/**
 * Your NumMatrix object will be instantiated and called as such:
 * NumMatrix obj = new NumMatrix(matrix);
 * int param_1 = obj.sumRegion(row1,col1,row2,col2);
 */

File: homework/hoanghai/lesson_1/304.RangeSumQuery2D-Immutable.js
/**
 * @param {number[][]} matrix
 */
 let sums
 var NumMatrix = function(matrix) {
     // matrix n row x m col
     let n = matrix.length;
     let m = matrix[0].length;
     sums = [...Array(n+1)].map(x => Array(m+1).fill(0))
     for (let i = 1; i <= n; i++) {
         for (let j = 1; j <= m; j++) {
             sums[i][j] = sums[i-1][j] + sums[i][j-1] + matrix[i-1][j-1] - sums[i-1][j-1];
         }
     }
 };
 
 /** 
  * @param {number} row1 
  * @param {number} col1 
  * @param {number} row2 
  * @param {number} col2
  * @return {number}
  */
 NumMatrix.prototype.sumRegion = function(row1, col1, row2, col2) {
     let outsideRec = sums[row2+1][col2+1];
     let inside2smallRec = sums[row1][col2+1] + sums[row2+1][col1];
     let overlapRec = sums[row1][col1];
     return outsideRec - inside2smallRec + overlapRec;
 };
 
 /** 
  * Your NumMatrix object will be instantiated and called as such:
  * var obj = new NumMatrix(matrix)
  * var param_1 = obj.sumRegion(row1,col1,row2,col2)
  */